Object Model to Transport Mapping (#333)
moq-wg/moq-transport: draft-ietf-moq-transport.md

(see {{message-subscribe-fin}}) or a SUBSCRIBE_RST
(see {{message-subscribe-rst}}).

A relay MUST not reorder or drop objects received on a multi-object stream when

> forwarding to subscribers.

Reviewer: I think we said in a previous author's call that we would relax this when we formally add concepts like TTL or drop preference, which give relay instructions about which objects are droppable.  I can update to "A relay MUST not reorder or drop objects *without application specific information*" - or something like that?

> forwarding to subscribers.

Author: works for me. Thanks

TODO: figure out how a relay closes these streams

When a stream begins with `STREAM_HEADER_TRACK`, all objects on the stream
belong to the track requested in the Subscribe message identified by `Subscribe
ID`.  All objects on the stream have the same `Object Send Order`.

~~~
STREAM_HEADER_TRACK Message {
  Subscribe ID (i)
  Object Send Order (i),
}
~~~
{: #stream-header-track-format title="MOQT STREAM_HEADER_TRACK Message"}

All Objects received on a stream opened with STREAM_HEADER_TRACK have an `Object
Forwarding Preference` = `Track`.

To send an Object with `Object Forwarding Preference` = `Track`, find the open
stream that is associated with the subscription, or open a new one and send the
`STREAM_HEADER_TRACK` if needed, then serialize the the following object fields.

~~~
{
  Group ID (i),
  Object ID (i),
  Object Payload Length (i),
  Object Payload (...),
}
~~~
{: #object-track-format title="MOQT Track Stream Object Fields"}

When a stream begins with `STREAM_HEADER_GROUP`, all objects on the stream
belong to the track requested in the Subscribe message identified by `Subscribe
ID` and the group indicated by `Group ID`.  All objects on the stream

> have the same `Object Send Order`.

Reviewer: The intent is to tell the receiver how to populate the canonical 'Object Send Order'.  I can say "have the `Object Send Order` specified in the stream header" ?

~~~
STREAM_HEADER_GROUP Message {
  Subscribe ID (i),
  Group ID (i)
  Object Send Order (i)
}
~~~
{: #stream-header-group-format title="MOQT STREAM_HEADER_GROUP Message"}

All Objects received on a stream opened with `STREAM_HEADER_GROUP` have an
`Object Forwarding Preference` = `Group`.

To send an Object with `Object Forwarding Preference` = `Group`, find the open
stream that is associated with the subscription, `Group ID` and `Object

> Send Order`, or open a new one and send the `STREAM_HEADER_GROUP` if needed,

Author: I really don't see a need to have a match on the SendOrder here. StreamPerGroup sends all the objects with the same groupId. if the application wants to send each priority in its own group, then it arranges that way.

Reviewer: The intent was to allow more flexibility than strictly stream-per-group.  With this design, the publisher can split a group into multiple streams with different priorities.  The original proposal 3 had both kinds of streams and the feedback was to combine them.

then serialize the following fields.

~~~
{
  Object ID (i),
  Object Payload Length (i),
  Object Payload (...),
}
~~~
{: #object-group-format title="MOQT Group Stream Object Fields"}

### Examples:

Sending a track on one stream:

~~~
STREAM_HEADER_TRACK {
  Subscribe ID = 1
  Track Alias = 1
  Object Send Order = 0
}
{
  Group ID = 0
  Object ID = 0
  Object Payload Length = 4
  Payload = "abcd"
}
{
  Group Sequence = 1

>   Object Sequence = 0

Author: sequence to Id

  Object Payload Length = 4
  Payload = "efgh"
}

> ~~~

Author: May be we should have this example to extended as 
{Stream_Header_Track}{g1, o0}{g1, o1} ... {g2,o0}, {g2,o1}, ....

Sending a group on one stream, with a unordered object in the group appearing
on its own stream.

> 

Author: the unordered object example is confusing here and I prefer not to include and create a new example for OBJECT_STREAM

~~~
Stream = 2

STREAM_HEADER_GROUP {
  Subscribe ID = 2
  Track Alias = 2

>   Group Sequence = 0

Author: sequence to Id

serialize object fields below, and terminate the stream.

~~~
OBJECT_STREAM Message {
  Subscribe ID (i),
  Track Alias (i),
  Group ID (i),
  Object ID (i),
  Object Send Order (i),
  Object Payload (...),
}
~~~
{: #moq-transport-object-stream-format title="MOQT OBJECT_STREAM Message"}

* Subscribe ID: Subscription Identifer as defined in {{message-subscribe-req}}.

* Track Alias: Identifies the Track Namespace and Track Name as defined in
{{message-subscribe-req}}.

The Track Namespace and Track Name identified by the Track Alias is different
from the one specified in the subscription identified by Subscribe ID, the
receiver MUST close the session with a Protocol Violation.

* Other fields: As described in {{canonical-object-fields}}.

### Multi-Object Streams

When multiple objects are sent on a stream, the stream begins with a stream
header message and is followed by one or more sets of serialized object fields.
If a stream ends gracefully in the middle of a serialized Object, terminate the
session with a Protocol Violation.

A sender SHOULD NOT open more than one multi-object stream at a time with the
same stream header message type and fields.

> TODO: figure out how a relay closes these streams

Author: agree we need to get to this , but I feel this can be done in a separate PR>

> TODO: figure out how a relay closes these streams

Reviewer: > I don't think a relay can implement the draft until this is specified.

I agree.

> agree we need to get to this , but I feel this can be done in a separate PR

I'm inclined to take the consensus we have, merge it and build on it, rather than keeping this PR open.

I don't think we want to take it from the QUIC layer and we do need an MoQ layer signal.  Is the stream carrying the track data even sufficient or could it be reset?  The stream termination messages may need to go on the control stream?

When a stream begins with `STREAM_HEADER_TRACK`, all objects on the stream
belong to the track requested in the Subscribe message identified by `Subscribe
ID`.  All objects on the stream have the `Object Send Order` specified in the
stream header.

~~~
STREAM_HEADER_TRACK Message {
  Subscribe ID (i)
  Track Alias (i),
  Object Send Order (i),
}
~~~
{: #stream-header-track-format title="MOQT STREAM_HEADER_TRACK Message"}

All Objects received on a stream opened with STREAM_HEADER_TRACK have an `Object
Forwarding Preference` = `Track`.

To send an Object with `Object Forwarding Preference` = `Track`, find the open

> stream that is associated with the subscription, or open a new one and send the

Reviewer 2: I'm confused by this sentence. This seems like an implementation detail depending on the API?

Personally, I wouldn't lazy initialize the stream because then the MoQ library won't know when to close it. I think you can safely remove this sentence and just say that multiple objects follow.

Author: i think this needs to caoture the intent as " all the objects for a given track is sent on the same stream" .

Reviewer: The intent was to explain how relays send outgoing objects, but make it generic for publishers too.  Similar language was used for other multi-object stream formats.

I don't think it conflicts with your plan to not-lazy-initialize, that would fit in "find the open stream...", but I'm open to different language too.

`STREAM_HEADER_TRACK` if needed, then serialize the the following object fields.

~~~
{
  Group ID (i),

>   Object ID (i),

Reviewer 2: Can we add the constraint that both of these must be increasing over the life of the stream? Otherwise dropping and subsequent gap detection becomes very difficult.

Author: groupId and object Id are monotonically increasing integers.

Reviewer: I'm ok adding an explicit restriction here.  Does anyone object?

Reviewer 2: > groupId and object Id are monotonically increasing integers.

Are they monotonically increasing? I thought the use-case was two streams with objects interlaced, in which case they wouldn't be.

  Object Payload Length (i),
  Object Payload (...),
}
~~~
{: #object-track-format title="MOQT Track Stream Object Fields"}

When a stream begins with `STREAM_HEADER_GROUP`, all objects on the stream
belong to the track requested in the Subscribe message identified by `Subscribe
ID` and the group indicated by `Group ID`.  All objects on the stream
have the `Object Send Order` specified in the stream header.

~~~
STREAM_HEADER_GROUP Message {
  Subscribe ID (i),
  Track Alias (i),
  Group ID (i)
  Object Send Order (i)
}
~~~
{: #stream-header-group-format title="MOQT STREAM_HEADER_GROUP Message"}

All Objects received on a stream opened with `STREAM_HEADER_GROUP` have an
`Object Forwarding Preference` = `Group`.

To send an Object with `Object Forwarding Preference` = `Group`, find the open
stream that is associated with the subscription, `Group ID` and `Object
Send Order`, or open a new one and send the `STREAM_HEADER_GROUP` if needed,
then serialize the following fields.

~~~
{

>   Object ID (i),

Reviewer 2: Same thing, can we make this increasing only?

Reviewer: I'm ok adding an explicit restriction here.  Does anyone object?

  Object Payload Length (i),
  Object Payload (...),
}
~~~
{: #object-group-format title="MOQT Group Stream Object Fields"}

### Examples:

Sending a track on one stream:

~~~
STREAM_HEADER_TRACK {
  Subscribe ID = 1
  Track Alias = 1
  Object Send Order = 0
}
{
  Group ID = 0
  Object ID = 0
  Object Payload Length = 4
  Payload = "abcd"
}
{
  Group ID = 1
  Object ID = 0
  Object Payload Length = 4
  Payload = "efgh"
}
~~~

Sending a group on one stream, with a unordered object in the group appearing
on its own stream.

~~~
Stream = 2

STREAM_HEADER_GROUP {
  Subscribe ID = 2
  Track Alias = 2
  Group ID = 0
  Object Send Order = 0
}
{
  Object ID = 0
  Object Payload Length = 4
  Payload = "abcd"
}
{

>   Object ID = 1

Reviewer: Yes, I think I meant for this to be Object ID=2.

object. The preferences are Track, Group, and Object.  An Object MUST be sent
according to its `Object Forwarding Preference`, described below.

* Object Payload: An opaque payload intended for the consumer and SHOULD
NOT be processed by a relay.

### Object Message Formats

An `OBJECT_STREAM` message carries a single object on a stream.  There is no
explicit length of the payload; it is determined by the end of the stream.  An
`OBJECT_STREAM` message MUST be the first and only message on a unidirectional
stream.

An Object received in an `OBJECT_STREAM` message has an `Object Forwarding
Preference` = `Object`.

To send an Object with `Object Forwarding Preference` = `Object`, open a stream,
serialize object fields below, and terminate the stream.

~~~
OBJECT_STREAM Message {
  Subscribe ID (i),
  Track Alias (i),
  Group ID (i),
  Object ID (i),
  Object Send Order (i),
  Object Payload (...),
}
~~~
{: #moq-transport-object-stream-format title="MOQT OBJECT_STREAM Message"}

* Subscribe ID: Subscription Identifer as defined in {{message-subscribe-req}}.

* Track Alias: Identifies the Track Namespace and Track Name as defined in
{{message-subscribe-req}}.

> The Track Namespace and Track Name identified by the Track Alias is different

Reviewer 3: ```suggestion
If the Track Namespace and Track Name identified by the Track Alias is different
```

Reviewer 3: Also, this applies to all types of mappings, so maybe move it up to the end of the previous section?

The mode `StreamPerPriority` specifies that objects marked with a
given priorirty shared the same underlying QUIC stream. `StreamPerTrack` has a
value of 0x3.

##### Mode: Datagram

The mode `Datagram` specifies that objects are sent using QUIC datagrams and
has a value of 0x4.

Note: Certain considerations apply when using Datagram. See the
section on datagram that follows from discussions in the PR316.s
under the PR 316

TODO: More details are needed to determine when a stream can be closed.
      Refer to disucssions on issue 318 on explicit markers for signaling
      stream completion.

## OBJECT (Proposal 3)

### Canonical Object Fields

A canonical MoQ Object has the following information:

* Track Alias: The track alias obtained as part of subscription and/or
publish control message exchanges.

* Group Sequence : The object is a member of the indicated group
{{model-group}} within the track.

* Object Sequence: The order of the object within the group.  The
sequence starts at 0, increasing sequentially for each object within the
group.

* Object Send Order: An integer indicating the object send order
{{send-order}} or priority {{ordering-by-priorities}} value.

> * Object Forwarding Preference: An enumeration indicating how a sender sends an

Reviewer 4: We need a way to tell the publisher how to map QUIC streams and this seems like a good proposal to me. Perhaps we should add something about relay implementation, perhaps: The relay would need remember how those objects were received on the subscriber side and replicate that in the publisher side

Reviewer: I tried to capture this in the sections below describing how to determine the preference when receiving an object and how to send an object with a given preference.  I'm also going to add a normative bit to the relay section.

object. The preferences are Track, Group, Object, Priority and Datagram.  An

> Object MUST be sent according to its `Object Forwarding Preference`, described

Reviewer 4: If the "Object *MUST* be sent according to its `Object Forwarding Preference`", should we remove the word "Preference". It seems contradictory with MUST

Reviewer: Do we like "Object forwarding mode" better than preference?  I'm neutral.

below.

* Object Payload: An opaque payload intended for the consumer and SHOULD
NOT be processed by a relay.

### Object Message Formats

> An `OBJECT_STREAM` message carries a single object on a stream.  There is no

Author: s/on a stream/on a QUIC stream

Reviewer: I left out QUIC to account for WebTransport or QUIC.  I think this matches elsewhere now?  In an earlier draft I said "transport stream".

> explicit length of the payload; it is determined by the end of the stream.  An

Reviewer 5: I would actually be in favor of removing this message type.

I'm using objects with an unbounded length to get around the lack of stream mapping in the current draft. The idea is that an OBJECT_STREAM can't be moved or combined with other streams, so it's a way to actually enforce stream mapping across relays.

However we don't need it any longer if there's explicit stream mapping and ordering/reliability. I would be fine with each OBJECT having a specified length (like HTTP/3 DATA frames) provided the relay can't rearrange them. It does mean that we have to buffer each frame to determine the length but that's often unavoidable.

Reviewer: > A relay could still convert single object streams into multi object streams that by buffering until EOF.

I don't think so.  The rules for sending objects say

```
To send an Object with `Object Forwarding Preference` = `Object`, open a stream,
serialize object fields below, and terminate the stream.
```

> `OBJECT_STREAM` message MUST be the first message on a unidirectional stream.

Author: I think we should also say , there MUST be only occurrence of `OBJECT_STREAM` message else its an protocol error.

Reviewer: Is that not implied by saying that the payload length is determined by the length of the stream?

> `OBJECT_STREAM` message MUST be the first message on a unidirectional stream.

Reviewer 3: ```suggestion
`OBJECT_STREAM` message MUST be the first and only message on a unidirectional stream.
```

An Object received in an `OBJECT_STREAM` message has an `Object Forwarding
Preference` = `Object`.

To send an Object with `Object Forwarding Preference` = `Object`, open a stream,
serialize object fields below, and terminate the stream.

~~~

> OBJECT_STREAM Message {

Reviewer 3: Don't we need a type value so if we receive a stream, we know what type of stream it is?

Reviewer: Jordi is right -- it's the same.

  Track Alias (i),
  Group Sequence (i),
  Object Sequence (i),
  Object Send Order (i),
  Object Payload (...),
}
~~~
{: #moq-transport-object-stream-format title="MOQT OBJECT_STREAM Message"}

An `OBJECT_DATAGRAM_PREFERRED` message carries a single object in a datagram or
a stream. There is no explicit length of the payload; it is determined by the
length of the datagram or stream.  If this message appears on a stream, it MUST
be the first message on a unidirectional stream.

An Object received in an `OBJECT_DATAGRAM_PREFERRED` message has an `Object
Forwarding Preference` = `Datagram`.

> To send an Object with `Object Forwarding Preference` = `Datagram`, serialize
> the fields and payload, and compare the serialized size with the maximum

Reviewer 3: ```suggestion
To send an Object with `Object Forwarding Preference` = `Datagram`, determine
the length of the fields and payload, and compare the length with the maximum
```

Reviewer: I'm going to remove datagram for now.

> datagram size of the session.  If the object size is less than or equal maximum
> datagram size, send the serialized data as a datagram.  Otherwise, open a
> stream, send the serialized data and terminate the stream.  An implementation

Reviewer 3: ```suggestion
datagram size of the session.  If the object size is less than or equal maximum
datagram size, send the serialized data as a datagram.  Otherwise, the object on a
unidirectional stream as you would an OBJECT_STREAM message.  An implementation
```

Reviewer: same as above.

> stream, send the serialized data and terminate the stream.  An implementation

Reviewer 5: We do need to send RESET_STREAM for flow control reasons. What error code do we use? Do we need to indicate why the datagram was dropped (ex. flow control)?

Does the receiver also reply with a STOP_SENDING?

MUST NOT send an Object with `Object Forwarding Preference` = `Datagram` on a
stream if it is possible to send it as a datagram.

~~~

> OBJECT_DATAGRAM_PREFERRED Message {

Reviewer 5: I think this is a footgun. It has the same downsides as IP fragmentation.

The application want specific behavior from datagrams, namely that they are independent. If a relay converts a datagram a stream, it will be split across IP packets and may even share packets with other datagram fragments. This makes implementing something like FEC basically impossible.

In my opinion, we should require the datagram extension and that all OBJECT_DATAGRAM messages are smaller than 1.2kB.

Reviewer 5: Also, is a relay expected to cache datagrams or does it require timely delivery? I feel like the intent of datagrams is for transient data, but currently it only disables retransmissions. That feels like the wrong behavior, especially with datagrams that span multiple MTU because it only serves to amplify loss.

  Track Alias (i),
  Group Sequence (i),
  Object Sequence (i),

>   Object Send Order (i),

Reviewer: This field is to convey the priority signal to the next hop, which might be using an API that allows setting priority on datagrams.

  Object Payload (...),
}
~~~
{: #object-datagram-format title="MOQT OBJECT_DATAGRAM_PREFERRED"}

### Multi-Object Streams

When multiple objects are sent on a stream, the stream begins with a stream
header message and is followed by one or more sets of serialized object fields.

> If a stream ends in the middle of a serialized Object, terminate the session

Reviewer 3: What if the stream is reset?

Reviewer: You can terminate it cleanly on an object boundary, or reset it, but you can't fin in the middle of an object.

with a Protocol Violation.

TODO: figure out how a relay closes these streams

When a stream begins with `STREAM_HEADER_TRACK`, all objects on the stream
belong to track indicated by `Track Alias`.

~~~

> STREAM_HEADER_TRACK Message {

Reviewer 4: I assume we can send 1 track (ex: video track) in mutiple Streams with STREAM_HEADER_TRACK, can I or that would be a protocol violation?

Reviewer: I'm going to add that a sender SHOULD only have one stream open with the exact same STREAM_HEADER.  It's not all that enforceable at the receiver though because of reordering.

  Track Alias (i)
}
~~~
{: #stream-header-track-format title="MOQT STREAM_HEADER_TRACK Message"}

> All Objects received on a stream opened with STREAM_HEADER_TRACK have an `Object

Reviewer 3: TRACK_STREAM or just TRACK would be simpler to me.

Forwarding Preference` = `Track`.

To send an Object with `Object Forwarding Preference` = `Track`, find the open
stream that is associated with the `Track Alias`, or open a new one and send the

> `STREAM_HEADER_TRACK` if needed, then serialize the the following object fields.

Reviewer 3: How do I know if they're needed?

Do we have one Object message at the beginning of the stream after the 'type' is serialized and then omit the duplicate fields on subsequent Object messages?

Or do we have a 'message' that starts the stream that could be only the type, followed by the shared values, and then a sequence of objects?  They're fairly equivalent, but we should pick one.

Reviewer: > How do I know if they're needed?

If you have to open a stream for this, then you need to send a header.

> Or do we have a 'message' that starts the stream that could be only the type, followed by the shared values, and then a sequence of objects?

This one.

~~~
{
  Group Sequence (i),
  Object Sequence (i),
  Object Send Order (i),
  Object Payload Length (i),
  Object Payload (...),
}
~~~
{: #object-track-format title="MOQT Track Stream Object Fields"}

> When a stream begins with `STREAM_HEADER_GROUP`, all objects on the stream

Reviewer 3: How about GROUP?  Also, it'd be more natural to put this above TRACK in terms of ordering.

Reviewer: Ok

belong to the track indicated by the `Track Alias` and the group indicated by
`Group Sequence`.

~~~
STREAM_HEADER_GROUP Message {
  Track Alias (i),
  Group Sequence (i)
}
~~~
{: #stream-header-group-format title="MOQT STREAM_HEADER_GROUP Message"}

All Objects received on a stream opened with `STREAM_HEADER_GROUP` have an
`Object Forwarding Preference` = `Group`.

To send an Object with `Object Forwarding Preference` = `Group`, find the open
stream that is associated with the `Track Alias` and `Group Sequence`, or open a
new one and send the `STREAM_HEADER_GROUP` if needed, then serialize the
following fields.

~~~
{
  Object Sequence (i),
  Object Send Order (i),
  Object Payload Length (i),
  Object Payload (...),
}
~~~
{: #object-group-format title="MOQT Group Stream Object Fields"}

When a stream begins with `STREAM_HEADER_PRIORITY`, all objects on the stream
belong have the same `Track Alias`, `Group Sequence` and `Object Send Order`.

~~~

> STREAM_HEADER_PRIORITY Message {

Reviewer 3: I don't think we need this.  I'd rather have every stream have a single priority.  If we want to do that later we can, but we all agree that mixing priorities on a single stream is problematic at best.

Reviewer: > I'd rather have every stream have a single priority. 

If that's the case, I think you want this message and remove the GROUP one that allows a different send order per object.

Reviewer 2: +1 priority should definitely be hoisted to the stream header.

There's an argument that a relay could perform application-level prioritization by reordering objects on a stream, but that suffers from head-of-line blocking and contradicts the ordering/reliability that QUIC streams are supposed to provide.

NOT be processed by a relay.

### Object Delivery Preferences {#delivery-prefs}

The publisher specifies its preference for mapping the elements of the
object model - Tracks, Groups and Objects, to the underlying transport
mechanisms such as QUIC Streams and QUIC Datagrams. A publisher chooses it
preference driven by the application needs.

> #### Proposal 1 - Stream Placement Indicator (Victor's Proposal and Text)

Author: Could you please clarify the major problems that you see ? also what do you see the concerns with different hops having different mappings ( not that I am proposing we allow that , but curious to understand more)

> #### Proposal 1 - Stream Placement Indicator (Victor's Proposal and Text)

Reviewer 2: As Cullen said at the mic, it's just a bug if the relay doesn't forward streams as they were intended.

Packet loss will delay the delivery of a stream relative to others. This algorithm is quite stateful and works based on the most recently received object (regardless of stream) so it will race like crazy and produce wildly unexpected results.

`Stream Placement Indicator` specifies as variable length integer informing
the stream to choose for sending a given object.

Each publisher follows the procedure defined below :

Let `N` be the object sequence of the current object and `D` be the
stream placement indicator, and `S` be a stack of streams that is initially
empty.

For each object to be sent:

1. If `D` is zero, open a new stream and push it at the top of the stack.
Send the object on the stream at the top of the stack.

2. If `D` is non-zero, check if the object sequence of the most recent object
that was sent on the stream at the top of the stack has a value of  `N - D`.

  -  If it is, continue sending the current object on the stream at the top
     of the stack.

  - If it is not, close the stream at the top of the stack, pop the stream
     off the stack and repeat from the Step 2. If the stack is empty, close
     the connection with appropriate error code.

> #### Proposal 2 - Object Placement Mode

Reviewer 2: It's so verbose to put this mode in every OBJECT. There's also a plethora of collisions that make it almost impossible to use multiple modes.

What about a single mode per track in SUBSCRIBE_OK instead?

`Object Placement Mode` is a enumeration of possible mappings of the object
model to the underlying transport mechanisms.

> ##### Mode: StreamPerGroup

Author: Stream per group, every object that belongs to Group 1, send it on the stream that is assigned for Group 1.  So where is the out of order scenario coming from ? 

Gaps has nothing to do with the transport mode mapping. A gap can be created due to object being dropped by a node on the path and if there exists an object for a group, it gets sent on the stream associated with the group

> ##### Mode: StreamPerGroup

Reviewer 2: > Stream per group, every object that belongs to Group 1, send it on the stream that is assigned for Group 1. So where is the out of order scenario coming from ?

If objects arrive on different streams then they can arrive out of intended order. If we force implicit == explicit then this isn't an issue, but the text should mention that.

> Gaps has nothing to do with the transport mode mapping. A gap can be created due to object being dropped by a node on the path and if there exists an object for a group, it gets sent on the stream associated with the group

That's fair, but then we need another mode with no gaps. An application that uses delta updates cannot tolerate an arbitrary hole in the middle.

> The mode `StreamPerGroup` specifies that all the objects for a given group

Reviewer 2: All OBJECTs in the specific group MUST also use mode StreamPerGroup.

Author: that's the intention with the existing statement

Reviewer 2: Yeah sorry I was suggesting that you add it.

are sent over a single QUIC stream. `StreamPerGroup` has a value of 0x0.

> ##### Mode: StreamPerObject

Reviewer 2: If I set this mode for an OBJECT that would then prohibit:

- any OBJECT in the group from using StreamPerGroup
- any OBJECT in the track from using StreamPerTrack
- any OBJECT in the priority from using StreamPerPriority

Author: If you want send all the objects for a track over single stream , you set the mode as StreamPerTrack,
if you want to send all the objects in a group together, you set StreamPerGroup,
if you want to send objects on independent streams, you mark them as StreamPerObject.

Reviewer 2: To clarify, I'm talking about when OBJECTs could have different modes. If you want ALL objects to use the same mode then that shouldn't be encoded per-OBJECT, it could be in SUBSCRIBE_OK instead.

Reviewer 6: Binding the transport mode to the subscription seems like a reasonable constraint. Once a publisher is producing a track, I can't find an obvious use-case as to why they would need to alter the object-to-transport mapping. If there was one in the future, the publisher could work around it by terminating the track and advertising a new one with the new mapping. Binding the mapping to the subscription means, as kixelated points out, that the mapping info can be sent just once in SUSBCRIBE_OK. For compactness, anything that doesn't change over the life of a subscription (such as track alias, subscribe ID and now mapping) should be communicated via SUBSCRIBE_OK.

The mode `StreamPerObject` specifies that each object is sent
over a single QUIC stream. `StreamPerObject` has a value of 0x1.

##### Mode: StreamPerTrack

> The mode `StreamPerTrack` specifies that all the groups and objects for a

Reviewer 2: All OBJECTs in the specified track MUST also use mode StreamPerTrack

given track are sent over a single QUIC stream. `StreamPerTrack` has a
value of 0x2.

##### Mode: StreamPerPriority

> The mode `StreamPerPriority` specifies that objects marked with a

Author: that's right

> The mode `StreamPerPriority` specifies that objects marked with a

Reviewer 2: Sorry, is this a stream...

- per priority per track?
- per priority per namespace?
- per priority (per session)?

The last one is my interpretation of the text and it's impossible because some other namespace/track will definitely want to use another mode.

belong to the track requested in the Subscribe message identified by `Subscribe
ID`.  All objects on the stream have the same `Object Send Order`.

~~~
STREAM_HEADER_TRACK Message {
  Subscribe ID (i)
  Object Send Order (i),
}
~~~
{: #stream-header-track-format title="MOQT STREAM_HEADER_TRACK Message"}

All Objects received on a stream opened with STREAM_HEADER_TRACK have an `Object
Forwarding Preference` = `Track`.

To send an Object with `Object Forwarding Preference` = `Track`, find the open
stream that is associated with the subscription, or open a new one and send the
`STREAM_HEADER_TRACK` if needed, then serialize the the following object fields.

~~~
{
  Group Sequence (i),
  Object Sequence (i),
  Object Payload Length (i),
  Object Payload (...),
}
~~~
{: #object-track-format title="MOQT Track Stream Object Fields"}

When a stream begins with `STREAM_HEADER_GROUP`, all objects on the stream
belong to the track requested in the Subscribe message identified by `Subscribe
ID` and the group indicated by `Group Sequence`.  All objects on the stream
have the same `Object Send Order`.

~~~
STREAM_HEADER_GROUP Message {

>   Subscribe ID (i),

Reviewer 7: I think this should be track alias not subscribe ID
